metaflow/plugins/kubernetes/kubernetes_decorator.py

```diff
@@ -387,14 +403,59 @@ def _save_package_once(cls, flow_datastore, package):
                 [package.blob], len_hint=1
             )[0]
 
+    @classmethod
+    def _parse_decorator_spec(cls, deco_spec: str):
+        if not deco_spec:
+            return cls()
+
+        valid_options = "|".join(cls.defaults.keys())
+        deco_spec_parts = []
+        for part in re.split(f""",(?=[\s\w]+[{valid_options}]=)""", deco_spec):
+            name, val = part.split("=", 1)
+            if name in {"labels", "node_selector"}:
+                try:
+                    json.loads(val.strip().replace('\\"', '"'))
+                except json.JSONDecodeError:
+                    both = name == "node_selector"
+                    val = json.dumps(
+                        cls.parse_kube_keyvalue_list(val.split(","), both),
+                        separators=(",", ":"),
+                    )
+            deco_spec_parts.append("=".join([name, val]))
+        deco_spec_parsed = ",".join(deco_spec_parts)
+        return super()._parse_decorator_spec(deco_spec_parsed)
+
     @staticmethod
-    def parse_node_selector(node_selector: list):
+    def parse_kube_keyvalue_list(items: List[str], requires_both: bool = True):
         try:
-            return {
-                str(k.split("=", 1)[0]): str(k.split("=", 1)[1])
-                for k in node_selector or []
-            }
+            ret = {}
+            for item_str in items:
+                item = item_str.split("=", 1)
+                if requires_both:
+                    item[1]  # raise IndexError
+                ret[str(item[0])] = str(item[1]) if len(item) > 1 else None
+            return ret
         except (AttributeError, IndexError):
-            raise KubernetesException(
-                "Unable to parse node_selector: %s" % node_selector
-            )
+            raise KubernetesException("Unable to parse kubernetes list: %s" % items)
+
+    @staticmethod
+    def validate_kube_labels(
+        labels: Optional[Dict[str, Optional[str]]],
+        regex_match: str = r"^(([A-Za-z0-9][-A-Za-z0-9_.]{0,61})?[A-Za-z0-9])?$",
+    ):
+        def validate_label(s: Optional[str]):
+            if not s:
+                # allow empty label
+                return s
+            if not re.search(regex_match, s):
+                # this is the same message kubernetes itself returns
+                raise Exception(
+                    f'Invalid value: "{s}": a valid label must be an empty string or '
+                    "consist of alphanumeric characters, '-', '_' or '.', and must "
+                    "start and end with an alphanumeric character (e.g. 'MyValue', "
+                    "or 'my_value', or '12345', regex used for validation is "
+                    "'^(([A-Za-z0-9][-A-Za-z0-9_.]{0,61})?[A-Za-z0-9])?$'"
+                )
+            return s
+
+        return {k: validate_label(v) for k, v in labels.items()} if labels else labels
```

Comment:
The keys would need to pass a more complicated regex.  Basically, you can have a prefix for keys that is a valid DSN and can be up to 256 characters.  But also, one reason I am not validating the keys is that in theory the key is more static than the value.  I would just let bad keys fail and let the user get the kubernetes message.  If you want me to look for those reserved prefixes (`kubernetes.io` and `k8s.io`), I could easily do that.  I don't think it is worth it tho.

Reply:
Agree that it might become unnecessarily complex with the key validation, we can skip this for the time being then. Maybe good to write a comment in the code for future developers as to why we only validate the values.